io-sim/test/Test/STM.hs

```diff
@@ -0,0 +1,795 @@
+      case nf1 of
+        NfReturn v -> evalTerm env' heap' allocs' t2
+          where
+            env' = extendEnv name v env
+        NfThrow v  -> (NfThrow v, heap', allocs')
+        NfRetry    -> (NfRetry,   heap', allocs')
+
+    OrElse t1 t2 ->
+      let (nft1, heap', allocs') = evalTerm env heap allocs t1 in
+      case nft1 of
+        NfReturn v -> (NfReturn v, heap', allocs')
+        NfThrow  v -> (NfThrow  v, heap', allocs')
+        NfRetry    -> evalTerm env heap allocs t2
+
+-- | The top level rule for STM transitions (on closed terms).
+--
+evalAtomically :: Term t -> (TxResult, Heap)
+evalAtomically t =
+    let env                  = Env mempty
+        heap                 = mempty
+        allocs               = mempty
+        (t', heap', allocs') = evalTerm env heap allocs t in
+    case t' of
+      NfReturn v -> (TxComitted v', heap')            -- ARET
+                      where v' = snapshotValue heap' v
+
+      NfThrow  v -> (TxAborted  v', heap <> allocs')  -- ATHROW
+                      where v' = snapshotValue heap' v
+
+      NfRetry    -> (TxBlocked, heap)  -- heap unchanged for retry.
+
+-- | Capture an immutable snapshot of a value, given the current value of the
+-- mutable variable heap.
+--
+snapshotValue :: Heap -> Value t -> ImmValue
+snapshotValue _  ValUnit   = ImmValUnit
+snapshotValue _ (ValInt x) = ImmValInt x
+snapshotValue h (ValVar n) = ImmValVar (snapshotValue h (readVar h n))
+
+
+--
+-- Execution in an STM monad (real or sim)
+--
+
+data ExecValue m (t :: Type) where
+
+    ExecValUnit ::                           ExecValue m TyUnit
+    ExecValInt  :: Int                    -> ExecValue m TyInt
+    ExecValVar  :: LazyTVar m (ExecValue m t) 
+                -> TyRep t                -> ExecValue m (TyVar t)
+
+instance Show (ExecValue m t) where
+  show  ExecValUnit         = "ExecValUnit"
+  show (ExecValInt x)       = "ExecValInt " ++ show x
+  show (ExecValVar _ tyrep) = "ExecValVar (<tvar> :: " ++ show tyrep ++ ")"
+
+
+data SomeExecValue m where
+     SomeExecValue :: ExecValue m t -> SomeExecValue m
+
+deriving instance Show (SomeExecValue m)
+
+
+newtype ExecEnv m = ExecEnv (Map NameId (SomeExecValue m))
+  deriving (Semigroup, Monoid)
+
+tyRepExecValue :: ExecValue m t -> TyRep t
+tyRepExecValue  ExecValUnit         = TyRepUnit
+tyRepExecValue (ExecValInt _)       = TyRepInt
+tyRepExecValue (ExecValVar _ tyrep) = TyRepVar tyrep
+
+lookupExecEnv :: ExecEnv m -> Name t -> ExecValue m t
+lookupExecEnv (ExecEnv env) (Name name tyrep) =
+    fromMaybe (error "lookupExecEnv: no such var") $ do
+    SomeExecValue v <- Map.lookup name env
+    Refl            <- tyrep `eqTyRep` tyRepExecValue v
+    return v
+
+extendExecEnv :: Name t -> ExecValue m t -> ExecEnv m -> ExecEnv m
+extendExecEnv (Name name _tyrep) v (ExecEnv env) =
+    ExecEnv (Map.insert name (SomeExecValue v) env)
+
+
+-- | Execute an STM 'Term' in the 'STM' monad.
+--
+execTerm :: (MonadSTM m, MonadThrow (STM m))
```

Comment:
The name is slightly misleading, this does not execute it just interprets a term as an `STM` expression.

Reply:
I think it is executing. This function turns the STM term into an STM action that interprets the term and returns the STM term execution result. If you run this in "real" STM then it's very much executing. If you run it in the sim then sure it turns it into a deep embedding for later execution, but that's just a special feature of that instance of MonadSTM.